flink-table/flink-table-runtime-blink/src/test/java/org/apache/flink/table/runtime/operators/join/AsyncLookupJoinRunnerTest.java

```diff
@@ -0,0 +1,156 @@
+import org.apache.flink.table.data.GenericRowData;
+import org.apache.flink.table.data.RowData;
+import org.apache.flink.table.runtime.collector.TableFunctionCollector;
+import org.apache.flink.table.runtime.collector.TableFunctionResultFuture;
+import org.apache.flink.table.runtime.generated.GeneratedFunctionWrapper;
+import org.apache.flink.table.runtime.generated.GeneratedResultFutureWrapper;
+import org.apache.flink.table.runtime.operators.join.lookup.AsyncLookupJoinRunner;
+import org.apache.flink.table.runtime.typeutils.InternalTypeInfo;
+
+import org.junit.Test;
+
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.function.Supplier;
+
+import static org.apache.flink.table.data.StringData.fromString;
+import static org.apache.flink.table.runtime.util.StreamRecordUtils.row;
+import static org.junit.Assert.assertNotNull;
+import static org.junit.Assert.assertNull;
+import static org.junit.Assert.fail;
+
+/**
+ * Tests for {@link AsyncLookupJoinRunner}.
+ */
+public class AsyncLookupJoinRunnerTest {
+
+	@Test
+	public void testCloseAsyncLookupJoinRunner() throws Exception {
+		final InternalTypeInfo<RowData> rightRowTypeInfo = InternalTypeInfo.ofFields(
+				DataTypes.INT().getLogicalType(),
+				DataTypes.STRING().getLogicalType());
+		final AsyncLookupJoinRunner joinRunner = new AsyncLookupJoinRunner(
+				new GeneratedFunctionWrapper(new TestingFetcherFunction()),
+				new GeneratedResultFutureWrapper<>(new TestingFetcherResultFuture()),
+				rightRowTypeInfo,
+				rightRowTypeInfo,
+				true,
+				100);
+		assertNull(joinRunner.getAllResultFutures());
+		closeAsyncLookupJoinRunner(joinRunner);
+
+		joinRunner.setRuntimeContext(new MockStreamingRuntimeContext(false, 1, 0));
+		joinRunner.open(new Configuration());
+		assertNotNull(joinRunner.getAllResultFutures());
+		closeAsyncLookupJoinRunner(joinRunner);
+
+		joinRunner.open(new Configuration());
+		joinRunner.asyncInvoke(row(1, "a"), new TestingFetcherResultFuture());
+		assertNotNull(joinRunner.getAllResultFutures());
+		closeAsyncLookupJoinRunner(joinRunner);
+	}
+
+	private void closeAsyncLookupJoinRunner(AsyncLookupJoinRunner joinRunner) throws Exception {
+		try {
+			joinRunner.close();
+		} catch (NullPointerException e) {
+			fail("Expected close to fail with null pointer exception.");
+		}
+	}
+
+	// ---------------------------------------------------------------------------------
+
+	/**
+	 * The {@link TestingFetcherFunction} only accepts a single integer lookup key and
+	 * returns zero or one or more RowData.
+	 */
+	public static final class TestingFetcherFunction
+			extends AbstractRichFunction
+			implements AsyncFunction<RowData, RowData> {
+
+		private static final long serialVersionUID = 4018474964018227081L;
```

Comment:
looks like `TestingFetcherFunction` and  `TestingFetcherResultFuture` are redundant,  we defined same inner class in three tests with same versionId, I think it's an existed issue. Could we only define once and reuse them?

Reply:
> Could

@leonardBang I thought that it could be defined once and reused `TestingFetcherFunction` and `TestingFetcherResultFuture`. I would like to reuse them.

Reply:
> looks like `TestingFetcherFunction` and `TestingFetcherResultFuture` are redundant, we defined same inner class in three tests with same versionId, I think it's an existed issue. Could we only define once and reuse them?

@leonardBang I thought that it could be defined once and reused TestingFetcherFunction and TestingFetcherResultFuture. I would like to reuse them.